src/app/InvokeInteraction.h

```diff
@@ -0,0 +1,438 @@
+ * calling conventions:
+ *      1. Handle the request and if a response needs to be generated, call AddResponse on the passed in InvokeResponder object.
+ *      2. Handle the request and if no responses needs to to be generated, do nothing.
+ *      3. Handle the request if a response needs to be generated asynchronously, call IncrementHoldOffRef() on the passed in InvokeResponder object. This
+ *         is a signal to the responder object to hold-off sending any queued up responoses till a matching DecrementHoldOffRef() call is made by that handler.
+ *
+ */
+class InvokeResponder
+{
+public:
+    /*
+     * @brief
+     *
+     * This interface defines the expectations for a handler of command requests.
+     *
+     */
+    class CommandHandler
+    {
+    public:
+        /*
+         * @brief
+         *
+         * Called to handle a data element present within an Invoke Request message.
+         *
+         * @param invokeInteraction         A reference to the invoke interaction instance that corresponds to the invoke request being processed.
+         * @param commandParams             A reference to an object containing the parameters of the command request.
+         * @param payload                   If a data payload is present, a TLV reader positioned at that payload is provided. The reader 
+         *                                  shall not be assumed to exist past this call, so the application should save off the data if they desire to.
+         */
+        virtual CHIP_ERROR OnInvokeRequest(CommandParams &commandParams, InvokeResponder &invokeInteraction, TLV::TLVReader *payload) = 0;
+
+        virtual ~CommandHandler() {} 
+    };
+
+    /*
+     * @brief
+     *
+     * Initializes the interaction object with the exchange context associated with the received message
+     * as well as a rvalue reference to the packet buffer containing the message.
+     *
+     */
+    CHIP_ERROR Init(Messaging::ExchangeContext *apContext, System::PacketBufferHandle &&aBufferHandle);
+   
+    /*
+     * @brief
+     *
+     * Encode a response to a request by providing command parameters as well as a pointer to an encodable object.
+     * The object's contents are immediately serialized into a response packet buffer. 
+     *
+     */
+    CHIP_ERROR AddResponse(CommandParams aParams, IEncodableElement *serializable);
+   
+    /*
+     * @brief
+     *
+     * Encode a response to a request by providing command parameters as well as a lambda that is in turn
+     * invoked with a TLVWriter passed in positioned at the appropriate offset into the packet buffer.
+     *
+     * This uses SFINAE to ensure selection of the right overload when a closure is passed in vs. a pointer to
+     * a serializable object.
+     *
+     */
+    template <typename F, typename std::enable_if<std::is_class<F>::value, int>::type = 0>
+    CHIP_ERROR AddResponse(CommandParams aParams, F f) {
```

Comment:
Sync Notes: Instead of passing in a templated lambda, pass in the object directly to be serialized and call the templated `Encode` method that will encode that object correctly based on its type.